pg_idkit/pg_idkit--0.0.3.sql

```diff
@@ -0,0 +1,303 @@
+ * which only the ASCII letters “A” through “Z” are treated as letters, 
+ * and sorting is done strictly by character code byte values."
+ * Source: https://www.postgresql.org/docs/current/collation.html
+ *
+ * Reference implementation: https://github.com/segmentio/ksuid
+ * Also read: https://segment.com/blog/a-brief-history-of-the-uuid/
+ *
+ * MIT License.
+ */
+create or replace function gen_random_ksuid_microsecond() returns text as $$
+declare
+	v_time timestamp with time zone := null;
+	v_seconds numeric(50) := null;
+	v_micros numeric(50)  := null;
+	v_numeric numeric(50) := null;
+	v_epoch numeric(50) = 1400000000; -- 2014-05-13T16:53:20Z
+	v_base62 text := '';
+	v_alphabet char array[62] := array[
+		'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+		'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
+		'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 
+		'U', 'V', 'W', 'X', 'Y', 'Z', 
+		'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 
+		'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't',
+		'u', 'v', 'w', 'x', 'y', 'z'];
+	i integer := 0;
+begin
+
+	-- Get the current time
+	v_time := clock_timestamp();
+
+	-- Extract the epoch seconds and microseconds
+	v_seconds := EXTRACT(EPOCH FROM v_time) - v_epoch;
+	v_micros  := MOD((EXTRACT(microseconds FROM v_time)::numeric(50)), 1e6::numeric(50));
+
+	-- Generate a KSUID in a numeric variable
+	v_numeric := (v_seconds * pow(2::numeric(50), 128))  -- 32 bits for seconds
+		+ (v_micros * pow(2::numeric(50), 108))          -- 20 bits for microseconds and 108 bits for randomness
+		+ ((random()::numeric(70,20) * pow(2::numeric(70,20), 54))::numeric(50) * pow(2::numeric(50), 54)::numeric(50))
+		+  (random()::numeric(70,20) * pow(2::numeric(70,20), 54))::numeric(50);
+
+	-- Encode it to base-62
+	while v_numeric <> 0 loop
+		v_base62 := v_base62 || v_alphabet[mod(v_numeric, 62) + 1];
+		v_numeric := div(v_numeric, 62);
+	end loop;
+	v_base62 := reverse(v_base62);
+	v_base62 := lpad(v_base62, 27, '0');
+
+	return v_base62;
+	
+end $$ language plpgsql;
+
+
+
+/**
+ * Converts a KSUID into a timestamp (UTC).
+ */
+CREATE OR REPLACE FUNCTION ksuid_to_timestamp(ksuid text) 
+RETURNS timestamp with time zone 
+AS $$
+DECLARE
+  v_epoch CONSTANT numeric := 1400000000; -- Custom epoch timestamp
+  v_base62 text := '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
+  v_numeric numeric := 0;
+  v_timestamp timestamp with time zone;
+  i integer;
+BEGIN
+  -- Convert base-62 part of the KSUID to numeric
+  FOR i IN 1..length(ksuid) LOOP
+    v_numeric := v_numeric * 62 + strpos(v_base62, substring(ksuid FROM i FOR 1)) - 1;
+  END LOOP;
+
+  -- Calculate the timestamp
+  -- Divide by 2^128 to get the seconds part, as the first 32 bits are for the timestamp
+  v_numeric := v_numeric / pow(2::numeric, 128);
+  -- Convert the seconds since custom epoch to the timestamp
+  v_timestamp := to_timestamp(v_epoch + v_numeric);
+
+  RETURN v_timestamp;
+END $$ LANGUAGE plpgsql;
```

Comment:
this is generated by ChatGPT but it's pretty cool:

```sql
select gen_random_ksuid_second();

-- 2XgMWSlFKAiIsvY80ANAf9WIkUv

SELECT ksuid_to_timestamp('2XgMWSlFKAiIsvY80ANAf9WIkUv');

-- 2023-11-03 21:18:20.61971+00
```

some potential improvements:

1. Make the function generic enough to detect the "type" of KSUID (second/microsecond)
2. Allow the user to request a timezone (default to UTC)

Reply:
the idea is that you don't need to store a `created_at`, since it's implicit from the ID